File: internal/model/csv.go
package model

const CSVExportWarnRows = 10_000

// MaxCSVExportRows is deprecated for full-table export (no hard cap). Kept for reference in docs/tests.
const MaxCSVExportRows = MaxQueryRows

type CSVFormatOptions struct {
	Delimiter        string `json:"delimiter"`
	QuoteChar        string `json:"quoteChar"`
	HasHeader        bool   `json:"hasHeader"`
	NullValue        string `json:"nullValue"`
	LineEnding       string `json:"lineEnding"`
	Encoding         string `json:"encoding"`
	CommentChar      string `json:"commentChar"`
	LazyQuotes       bool   `json:"lazyQuotes"`
	TrimLeadingSpace bool   `json:"trimLeadingSpace"`
}

func DefaultCSVFormat() CSVFormatOptions {
	return CSVFormatOptions{
		Delimiter:  ",",
		QuoteChar:  `"`,
		HasHeader:  true,
		LineEnding: "crlf",
		Encoding:   "utf-8",
	}
}

func (o CSVFormatOptions) Normalized() CSVFormatOptions {
	n := DefaultCSVFormat()
	if o.Delimiter != "" {
		n.Delimiter = o.Delimiter
	}
	if o.QuoteChar != "" {
		n.QuoteChar = o.QuoteChar
	}
	n.HasHeader = o.HasHeader
	if o.LineEnding != "" {
		n.LineEnding = o.LineEnding
	}
	if o.Encoding != "" {
		n.Encoding = o.Encoding
	}
	n.NullValue = o.NullValue
	n.CommentChar = o.CommentChar
	n.LazyQuotes = o.LazyQuotes
	n.TrimLeadingSpace = o.TrimLeadingSpace
	return n
}

type ExportTableCSVRequest struct {
	ConnectionID string           `json:"connectionId"`
	TableName    string           `json:"tableName"`
	Format       CSVFormatOptions `json:"format"`
	Columns      []string         `json:"columns,omitempty"` // empty = all table columns
	DefaultPath  string           `json:"defaultPath,omitempty"`
	ExportID     string           `json:"exportId,omitempty"`
}

type ExportProgress struct {
	ExportID string `json:"exportId"`
	Exported int    `json:"exported"`
}

type ParseCSVPreviewRequest struct {
	FilePath string           `json:"filePath"`
	MaxRows  int              `json:"maxRows"`
	Format   CSVFormatOptions `json:"format"`
}

type CSVPreview struct {
	Headers   []string         `json:"headers"`
	Rows      []map[string]any `json:"rows"`
	RowCount  int              `json:"rowCount"`
	HasHeader bool             `json:"hasHeader"`
}

type ImportColumnSpec struct {
	DataType   string `json:"dataType"`
	PrimaryKey bool   `json:"primaryKey"`
}

type ImportCSVRequest struct {
	ConnectionID    string                      `json:"connectionId"`
	TargetTable     string                      `json:"targetTable"`
	NewTableName    string                      `json:"newTableName"`
	Mode            string                      `json:"mode"`
	ColumnMap       map[string]string           `json:"columnMap"`
	NewTableColumns map[string]ImportColumnSpec `json:"newTableColumns,omitempty"`
	FilePath        string                      `json:"filePath"`
	UpsertKeys      []string                    `json:"upsertKeys"`
	Format          CSVFormatOptions            `json:"format"`
}

type ImportCSVResult struct {
	RowsInserted int `json:"rowsInserted"`
	RowsUpdated  int `json:"rowsUpdated"`
}

File: frontend/src/lib/types/csv.ts
export interface CSVFormatOptions {
  delimiter: string
  quoteChar: string
  hasHeader: boolean
  lineEnding: string
  encoding: string
  commentChar?: string
  lazyQuotes?: boolean
  trimLeadingSpace?: boolean
}

export interface FileFilter {
  displayName: string
  pattern: string
}

export const CSV_EXPORT_WARN_ROWS = 10_000

export interface ExportTableCSVRequest {
  connectionId: string
  tableName: string
  format: CSVFormatOptions
  columns?: string[]
  defaultPath?: string
  exportId?: string
}

export interface ParseCSVPreviewRequest {
  filePath: string
  maxRows: number
  format: CSVFormatOptions
}

export interface CSVPreview {
  headers: string[]
  rows: Record<string, unknown>[]
  rowCount: number
  hasHeader: boolean
}

export interface ImportCSVRequest {
  connectionId: string
  targetTable: string
  newTableName: string
  mode: 'append' | 'update'
  columnMap: Record<string, string>
  filePath: string
  upsertKeys: string[]
  format: CSVFormatOptions
}

export interface ImportCSVResult {
  rowsInserted: number
  rowsUpdated: number
}

export function defaultCSVFormat(): CSVFormatOptions {
  return {
    delimiter: ',',
    quoteChar: '"',
    hasHeader: true,
    lineEnding: 'crlf',
    encoding: 'utf-8',
    commentChar: '',
    lazyQuotes: false,
    trimLeadingSpace: false,
  }
}

export function loadCSVFormatPreference(): CSVFormatOptions {
  try {
    const raw = localStorage.getItem('data-nexus:csv-format')
    if (raw) return { ...defaultCSVFormat(), ...JSON.parse(raw) }
  } catch {
    /* ignore */
  }
  return defaultCSVFormat()
}

export function saveCSVFormatPreference(format: CSVFormatOptions) {
  localStorage.setItem('data-nexus:csv-format', JSON.stringify(format))
}
